src/ring_theory/unique_factorization_domain.lean

```diff
@@ -455,3 +455,103 @@ def unique_factorization_domain.to_gcd_domain
+  (h : (∀ {p}, prime p → p ∣ a → p ∣ b → is_unit p)) : ∀ {p}, p ∣ a → p ∣ b → is_unit p :=
+λ p, induction_on_prime p
+  (by { simp only [zero_dvd_iff], intros, contradiction })
+  (λ x hx _ _, hx)
+  (λ p q hp hq ih dvd_a dvd_b, is_unit.mul
+    (h hq (dvd_of_mul_right_dvd dvd_a) (dvd_of_mul_right_dvd dvd_b))
+    (ih (dvd_of_mul_left_dvd dvd_a) (dvd_of_mul_left_dvd dvd_b)))
+
+lemma left_dvd_or_dvd_right_of_dvd_prime_mul {a : R} :
+  ∀ {b p : R}, prime p → a ∣ p * b → p ∣ a ∨ a ∣ b :=
+begin
+  refine induction_on_prime a _ _ _,
+  { intros b p _ ha,
+    refine (eq_zero_or_eq_zero_of_mul_eq_zero (zero_dvd_iff.mp ha)).imp _ _;
+      rintro ⟨rfl⟩; refl },
+  { intros x x_is_unit b _ _ _,
+    exact or.inr (is_unit_iff_forall_dvd.mp x_is_unit b) },
+  { intros a q a_ne_zero hq ih b p hp qa_dvd,
+    cases (hq.div_or_div (dvd_of_mul_right_dvd qa_dvd)) with q_dvd_p q_dvd_b,
+    { left,
+      apply dvd_mul_of_dvd_left,
+      refine dvd_symm_of_irreducible (irreducible_of_prime _) (irreducible_of_prime _) _;
+        assumption },
+    { rcases q_dvd_b with ⟨b', rfl⟩,
+      rw mul_left_comm at qa_dvd,
+      refine (ih hp ((mul_dvd_mul_iff_left hq.ne_zero).mp qa_dvd)).imp _ _,
+      { exact λ h, dvd_mul_of_dvd_right h _ },
+      { exact mul_dvd_mul_left q } } }
+end
+
+/-- Euclid's lemma: if `a ∣ b * c` and `a` and `c` have no common prime factors, `a ∣ b`.
+Compare `is_coprime.dvd_of_dvd_mul_left`. -/
+lemma dvd_of_dvd_mul_left_of_no_prime_factors {a b c : R} (ha : a ≠ 0) :
+  (∀ {p}, prime p → p ∣ a → p ∣ c → is_unit p) → a ∣ b * c → a ∣ b :=
+begin
+  refine induction_on_prime c _ _ _,
+  { intro no_factors,
+    simp only [dvd_zero, mul_zero, forall_prop_of_true],
+    haveI := classical.prop_decidable,
+    exact is_unit_iff_forall_dvd.mp
+      (no_factors_of_no_prime_factors ha @no_factors (dvd_refl a) (dvd_zero a)) _ },
+  { rintros _ ⟨x, rfl⟩ _ a_dvd_bx,
+    apply dvd_mul_unit_iff.mp a_dvd_bx },
+  { intros c p hc hp ih no_factors a_dvd_bpc,
+    apply ih (λ q hq dvd_a dvd_c, no_factors hq dvd_a (dvd_mul_of_dvd_right dvd_c _)),
+    rw mul_left_comm at a_dvd_bpc,
+    refine or.resolve_left (left_dvd_or_dvd_right_of_dvd_prime_mul hp a_dvd_bpc) (λ h, _),
+    apply hp.not_unit,
+    exact no_factors hp h (dvd_mul_right p c) }
+end
+
+/-- Euclid's lemma: if `a ∣ b * c` and `a` and `b` have no common prime factors, `a ∣ c`.
+Compare `is_coprime.dvd_of_dvd_mul_right`. -/
+lemma dvd_of_dvd_mul_right_of_no_prime_factors {a b c : R} (ha : a ≠ 0)
+  (no_factors : ∀ {p}, prime p → p ∣ a → p ∣ b → is_unit p) : a ∣ b * c → a ∣ c :=
+by simpa [mul_comm b c] using dvd_of_dvd_mul_left_of_no_prime_factors ha @no_factors
+
+/-- If `a ≠ 0, b` are elements of a unique factorization domain, then dividing
+out their common factor `c'` gives `a'` and `b'` with no factors in common. -/
+lemma exists_reduced_factors : ∀ (a ≠ (0 : R)) b,
+  ∃ a' b' c', (∀ {p}, p ∣ a' → p ∣ b' → is_unit p) ∧ c' * a' = a ∧ c' * b' = b :=
```

Comment:
Should `∀ {p}, p ∣ a' → p ∣ b' → is_unit p` become a definition/structure?

Reply:
At least I suggest using a different letter, because `p` makes me think you are talking about primes, whereas you are talking about units... so maybe `u`?

Reply:
I decided on `d` for divisor because with the other suggested changes, they are not assumed to be units either.